tests/copy_repo/test_copy_repo.py

```diff
@@ -0,0 +1,411 @@
+from more_executors.futures import f_return
+from pubtools.pulplib import (
+    Client,
+    ContainerImageRepository,
+    ErratumUnit,
+    FakeController,
+    FileRepository,
+    FileUnit,
+    ModulemdUnit,
+    RpmUnit,
+    YumRepository,
+)
+
+import pubtools._pulp.tasks.copy_repo
+from pubtools._pulp.tasks.copy_repo import CopyRepo
+from pubtools._pulp.ud import UdCacheClient
+
+
+class FakeUdCache(object):
+    def __init__(self):
+        self.flushed_repos = []
+        self.flushed_products = []
+
+    def flush_repo(self, repo_id):
+        self.flushed_repos.append(repo_id)
+        return f_return()
+
+    def flush_product(self, product_id):
+        self.flushed_products.append(product_id)
+        return f_return()
+
+
+class FakeCopyRepo(CopyRepo):
+    """copy-repo with services overridden for test"""
+
+    def __init__(self, *args, **kwargs):
+        super(FakeCopyRepo, self).__init__(*args, **kwargs)
+        self.pulp_client_controller = FakeController()
+        self._udcache_client = FakeUdCache()
+
+    @property
+    def pulp_client(self):
+        # Super should give a Pulp client
+        assert isinstance(super(FakeCopyRepo, self).pulp_client, Client)
+        # But we'll substitute our own
+        return self.pulp_client_controller.client
+
+    @property
+    def udcache_client(self):
+        # Super may or may not give a UD client, depends on arguments
+        from_super = super(FakeCopyRepo, self).udcache_client
+        if from_super:
+            # If it did create one, it should be this
+            assert isinstance(from_super, UdCacheClient)
+
+        # We'll substitute our own, only if UD client is being used
+        return self._udcache_client if from_super else None
+
+
+def test_missing_repos(command_tester):
+    """Command fails when pointed at nonexistent repos."""
+    # This one test covers the entry point
+
+    command_tester.test(
+        lambda: pubtools._pulp.tasks.copy_repo.entry_point(FakeCopyRepo),
+        [
+            "test-copy-repo",
+            "--pulp-url",
+            "https://pulp.example.com/",
+            "repo1,repo2 repo3,repo4",
```

Comment:
This part looks a bit suspicious.

When a person invokes this command, the repo pairs are likely to come in as separate arguments, e.g. the list would look like:

```
            "test-copy-repo",
            "--pulp-url",
            "https://pulp.example.com/",
            "repo1,repo2",
            "repo3,repo4",
```

It only wouldn't be that way if, for example, the user ran it from a shell and explicitly quoted it like:

```
pubtools-pulp-copy-repo --pulp-url https://pulp.example.com/ "repo1,repo2 repo3,repo4"
```

...and I wouldn't expect them to do that (and they shouldn't have to).

If the code works correctly in both cases I think that's fine, but I'd say the test should focus on the more common case, which should be the case of repo1,repo2 and repo3,repo4 arriving as two separate arguments.

Reply:
I took a look at the implementation and I don't think it actually does support this way, so I think this test only works because it's the test for missing repos.  It seems like internally it'd be splitting this into "repo1", "repo2 repo3", "repo4".  The test may 'work' since "repo2 repo3" isn't an existing repo but it wouldn't be what you intended.

Speaking of which, while reading that code I also noted that it doesn't really detect if the user messes up the format of the argument, so I'll comment about that too.

Reply:
Ah, okay. Yeah, I wrote it for that case because that's what was presented in the Jira description. I'll make sure it's capable of both.